virtual_rainforest/models/hydrology/hydrology_model.py

```diff
@@ -801,3 +876,147 @@ def calculate_drainage_map(grid: Grid, elevation: np.ndarray) -> dict[int, list[
+        max_capacity
+        * (1 - np.exp(-canopy_density_factor * precipitation / max_capacity)),
+        nan=0.0,
+    )
+
+
+def update_soil_moisture(
+    soil_moisture: NDArray[np.float32],
+    vertical_flow: NDArray[np.float32],
+    evapotranspiration: NDArray[np.float32],
+    soil_moisture_capacity: NDArray[np.float32],
+    soil_moisture_residual: NDArray[np.float32],
+) -> NDArray[np.float32]:
+    """Update soil moisture profile.
+
+    This function calculates soil moisture for each layer by removing the vertical flow
+    of the current layer and adding it to the layer below. Additionally, the
+    evapotranspiration is removed from the second soil layer.
+
+    Args:
+        soil_moisture: soil moisture after infiltration and surface evaporation, [mm]
+        vertical_flow: vertical flow between all layers, [mm]
+        evapotranspiration: canopy evaporation, [mm]
+        soil_moisture_capacity: soil moisture capacity for each layer, [mm]
+        soil_moisture_residual: residual soil moisture for each layer, [mm]
+
+    Returns:
+        updated soil moisture profile, relative volumetric water content, dimensionless
+    """
+    # TODO this is currently not conserving water
+    top_soil_moisture = np.clip(
+        soil_moisture[0] - vertical_flow[0],
+        soil_moisture_residual[0],
+        soil_moisture_capacity[0],
+    )
+
+    root_soil_moisture = np.clip(
+        soil_moisture[1] + vertical_flow[0] - vertical_flow[1] - evapotranspiration,
+        soil_moisture_residual[1],
+        soil_moisture_capacity[1],
+    )
+
+    if len(vertical_flow) == 2:
+        soil_moisture_updated = np.stack((top_soil_moisture, root_soil_moisture))
+
+    elif len(vertical_flow) > 2:
+        lower_soil_moisture = [
+            np.clip(
+                (soil_moisture[i + 1] + vertical_flow[i] - vertical_flow[i + 1]),
+                soil_moisture_residual[i + 1],
+                soil_moisture_capacity[i + 1],
+            )
+            for sm, vf in zip(
+                soil_moisture[2:],
+                vertical_flow[2:],
+            )
+            for i in range(len(soil_moisture) - 2)
+        ]
+        soil_moisture_updated = np.concatenate(
+            ([top_soil_moisture], [root_soil_moisture], lower_soil_moisture)
+        )
+
+    return soil_moisture_updated
+
+
+def calculate_layer_thickness(
+    soil_layer_heights: NDArray[np.float32],
+    meters_to_mm: float,
+) -> NDArray[np.float32]:
+    """Calculate layer thickness from soil layer depth profile.
+
+    Args:
+        soil_layer_heights: soil layer heights, [m]
+        meters_to_mm: meter to millimeter conversion factor
+
+    Returns:
+        soil layer thickness, mm
+    """
+    return np.array(
+        [
+            (soil_layer_heights[i] - soil_layer_heights[i - 1]) * (-meters_to_mm)
+            if i > 0
+            else soil_layer_heights[0] * (-meters_to_mm)
+            for i in range(len(soil_layer_heights))
+        ],
```

Comment:
I think you can get away with not creating an intermediate list here:

```suggestion
        (
            (soil_layer_heights[i] - soil_layer_heights[i - 1]) * (-meters_to_mm)
            if i > 0
            else soil_layer_heights[0] * (-meters_to_mm)
            for i in range(len(soil_layer_heights))
        ),
```

Reply:
Alternatively:
 
```python
np.diff(np.insert(soil_layer_depths, 0, 0)) * -meters_to_mm
```